shacl12-rules/index.html

```diff
@@ -700,163 +655,292 @@ <h3>RDF Rules Syntax</h3>
+            for each <var>μ</var><sub>2</sub> in T:
+              if compatible(<var>μ</var><sub>1</sub>, <var>μ</var><sub>2</sub>)
+                <var>μ</var><sub>3</sub> = merge(<var>μ</var>1, <var>μ</var><sub>2</sub>)
+                add <var>μ</var><sub>3</sub> to SEQ1
+            endfor
+        endif
 
     if rElt is a condition expression F:
-        T1 = empty list
-        for each row in T:
-            if ( eval(F,B) )
-                add R to T1
+        SEQ1 = {}
+        for each solution <var>μ</var> in SEQ:
+            ## EBV = Effective boolean value.
+            ## TODO: Eval errors.
+            if ( EBV(eval(<var>μ</var>, F)) is true )
+                add <var>μ</var> to SEQ1
                 endif
-        endfor
-    T = T1
-    endif
+            endfor
+        endif
 
     if rElt is an assignment(V, expr)
-        T1 = empty list
-        for each row in T:
-            let x = eval(expr, row)
-            let R1 = R union (V, x)
-            add R1 to T1
+        SEQ1 = {}
+        for each solution S in SEQ:
+            let x = eval(expr, <var>μ</var>)
+            add(V, x) to S
+            add S to SEQ1
             endfor
-        T = T1
         endif
+
+     if SEQ1 is empty 
+        SEQ = {}
+        exit
+        endif
+
+      SEQ = SEQ1
+   
     endfor
 
 # Evaluate rule head
 let H = empty set
-for each R in T:
-    Let S = set of triples obtained by replacing variables
-            in the triple templates of the head with values from R
-    define [head(R)/row]
+for each <var>μ</var> in T:
+    Let S = {}
+    for each triple template TT in head
+        Let triple = subst(<var>μ</var>, TT)
+        Add triple to S
     H = H union S
     endfor
 
 result eval(R, G) is H
 
 </pre>
-        <p>Note that `H` may contain triples that are also in the data grapoh.</p>
+      <p>Note that `H` may contain triples that are also in the data grapoh.</p>
       </section>
 
       <section id="eval-rule-set">
         <h4>Evaluation of a Rule Set</h4>
 
+        <p class=ednote>Sketch</p>
+
 <pre class="algorithm">
-Let G be the input RDF graph.
-Let I be the set of triples generated by evaluation
+Let G0 be the input RDF graph.
+Let GI be the set of triples generated by evaluation
 Let RS be a rule set
-Let finished = false
-
-while !finished:
-    finished = true
-    foreach rule in RS
-        let GI = G union I
-        let X = eval(rule, GI)
-        let Y = those triples in X that are not in GI
-        if Y is not empty:
-            finished = false
-            endif
-        let I = Y union I
-        endfor
-    endwhile
-
-result is GI
+Apply stratification to RS
+Let L be the sequence of layers after stratification
+
+for each layer in L
+    Let finished = false
+    while !finished:
+        finished = true
+        foreach rule in RS
+            let GI = G0 &cup; GI
+            let X = eval(rule, GI)
+            let Y = those triples in X that are not in G0
```

Comment:
```suggestion
            let Y = those triples in X that are not in GI
```

this should be GI, otherwise this never terminates

Reply:
(it was also GI previously)